packages/babel-plugin-transform-react-jsx/src/transform-experimental.js

```diff
@@ -0,0 +1,815 @@
+      args: args,
+    };
+
+    if (opts.pre) {
+      opts.pre(state, file);
+    }
+
+    let attribs = [];
+    let key;
+    let source;
+    let self;
+
+    // for React.jsx, key, __source (dev), and __self (dev) is passed in as
+    // a separate argument rather than in the args object. We go through the
+    // props and filter out these three keywords so we can pass them in
+    // as separate arguments later
+    for (let i = 0; i < openingPath.node.attributes.length; i++) {
+      const attr = openingPath.node.attributes[i];
+      if (t.isJSXAttribute(attr) && t.isJSXIdentifier(attr.name)) {
+        if (attr.name.name === "key") {
+          key = convertAttribute(attr).value;
+        } else if (attr.name.name === "__source") {
+          source = convertAttribute(attr).value;
+        } else if (attr.name.name === "__self") {
+          self = convertAttribute(attr).value;
+        } else {
+          attribs.push(attr);
+        }
+      } else {
+        attribs.push(attr);
+      }
+    }
+
+    if (attribs.length || path.node.children.length) {
+      attribs = buildJSXOpeningElementAttributes(
+        attribs,
+        file,
+        path.node.children,
+      );
+    } else {
+      // attributes should never be null
+      attribs = t.objectExpression([]);
+    }
+
+    args.push(attribs);
+
+    if (!file.opts.development) {
+      if (key !== undefined) {
+        args.push(key);
+      }
+    } else {
+      // isStaticChildren, __source, and __self are only used in development
+      args.push(
+        key === undefined ? t.identifier("undefined") : key,
+        t.booleanLiteral(path.node.children.length > 1),
+        source === undefined ? t.identifier("undefined") : source,
+        self === undefined ? t.identifier("undefined") : self,
+      );
+    }
+
+    if (opts.post) {
+      opts.post(state, file);
+    }
+
+    return (
+      state.call ||
+      t.callExpression(
+        path.node.children.length > 1 ? state.staticCallee : state.callee,
+        args,
+      )
+    );
+  }
+
+  function isChildrenProp(prop) {
+    return (
+      t.isJSXAttribute(prop) &&
+      t.isJSXIdentifier(prop.name) &&
+      prop.name.name === "children"
+    );
+  }
+
+  // Builds props for React.jsx. This function adds children into the props
+  // and ensures that props is always an object
+  function buildJSXOpeningElementAttributes(attribs, file, children) {
+    let _props = [];
+    const objs = [];
+
+    // In order to avoid having duplicate "children" keys, we avoid
+    // pushing the "children" prop if we have actual children. However,
+    // the children prop may have side effects, so to be certain
+    // these side effects are evaluated, we add them to the following prop
+    // as a sequence expression to preserve order. So:
+    // <div children={x++} foo={y}>{child}</div> becomes
+    // React.jsx('div', {foo: (x++, y), children: child});
+    // duplicateChildren contains the extra children prop values
```

Comment:
We could consider throwing in the transform if there are both syntactic children and the `children` property.

If we don't want to, we don't need this custom logic anyway. This output has the same effect:
```js
React.jsx("div", {
  children: x++,
  foo: y,
  children: child,
});
```

Reply:
We actually did this at some point but it broke on a previous browser because duplicate keys caused an error, so we had to do it this way instead. I'll add that to the comment though to clarify.

Reply:
There's a period in ECMAScript's lifetime when duplicate keys wasn't allowed in strict mode. Unfortunately they are still around. Babel should probably have a transform to remove duplicate keys from any object literal.

We could error but that would make it harder to upgrade. Since we did run into this, we know people write this code today. Arguably it shouldn't error for the same reasons ECMAScript doesn't error on duplicate keys.

Reply:
Yeah, Babel already has that transform. If people need to support a browser which throws for duplicate keys, it's handled by preset-env.

Since we are going to enable object spread by default in the jsx transform, most people will have preset-env anyway.

Reply:
What is the transform accounts for duplicate keys? I guess since this is for Babel 8 it's probably okay to introduce this as a breaking change, since we're also doing things like defaulting to spread anyway. @sebmarkbage

Reply:
It's https://babeljs.io/docs/en/babel-plugin-transform-duplicate-keys.

You can see it in action here: https://babeljs.io/repl/#?browsers=ie%2011&build=&builtIns=false&spec=false&loose=false&code_lz=G4QwTgBCELwQ3gKAhAHgLggRgDTLZgEyIC-QA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=false&presets=env%2Cenv&prettier=false&targets=&version=7.8.6&externalPlugins=%40babel%2Fplugin-transform-async-to-generator%407.8.3

It produces a sub-optimal output, but:
- The only popular browser that needs it is IE ([these browsers support them](https://github.com/babel/babel/blob/8f281f578db68ef2092ce35b8db003bdb346a9e4/packages/babel-compat-data/data/plugins.json#L199-L208))
- If we want to produce a more optimized output by merging the first duplicate key into the following one, it should be done in `@babel/plugin-transform-duplicate-keys` and not here
- Duplicate keys are super rare anyway